core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java

```diff
@@ -0,0 +1,391 @@
+    // table names as uppercase
+    boolean exists = false;
+    DatabaseMetaData dbMeta = dbConnPool.run(Connection::getMetaData);
+    ResultSet tables = dbMeta.getTables(null, null, SQL_TABLE_NAME, null);
+    if (tables.next()) {
+      exists = true;
+    }
+    tables.close();
+    ResultSet tablesUpper = dbMeta.getTables(null, null, SQL_TABLE_NAME.toUpperCase(), null);
+    if (tablesUpper.next()) {
+      exists = true;
+    }
+    tablesUpper.close();
+    ResultSet tablesLower = dbMeta.getTables(null, null, SQL_TABLE_NAME.toLowerCase(), null);
+    if (tablesLower.next()) {
+      exists = true;
+    }
+    tablesLower.close();
+
+    // create table if not exits
+    if (!exists) {
+      dbConnPool.run(conn -> conn.prepareStatement(SQL_TABLE_DDL).execute());
+      LOG.debug("Created table {} to store iceberg tables!", SQL_TABLE_NAME);
+    }
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    return new JdbcTableOperations(dbConnPool, fileIO, catalogName, tableIdentifier);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier table) {
+    if (table.hasNamespace()) {
+      return SLASH.join(warehouseLocation, table.namespace().levels(), table.name());
+    }
+    return SLASH.join(warehouseLocation, table.name());
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    TableOperations ops = newTableOps(identifier);
+    TableMetadata lastMetadata = ops.current();
+    try {
+      PreparedStatement sql = dbConnPool.run(c -> c.prepareStatement(SQL_DELETE_TABLE));
+      sql.setString(1, catalogName);
+      sql.setString(2, JdbcUtil.namespaceToString(identifier.namespace()));
+      sql.setString(3, identifier.name());
+      int deletedRecords = sql.executeUpdate();
+
+      if (deletedRecords > 0) {
+        LOG.debug("Successfully dropped table {}.", identifier);
+      } else {
+        throw new NoSuchTableException("Cannot drop table %s! table not found in the catalog.", identifier);
+      }
+
+      if (purge && lastMetadata != null) {
+        CatalogUtil.dropTableData(ops.io(), lastMetadata);
+        LOG.info("Table {} data purged!", identifier);
+      }
+      return true;
+    } catch (SQLException | InterruptedException e) {
+      throw new UncheckedSQLException("Failed to drop table!", e);
+    }
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    if (!this.namespaceExists(namespace)) {
+      throw new NoSuchNamespaceException("Namespace %s does not exist!", namespace);
+    }
+    try {
+      List<TableIdentifier> results = Lists.newArrayList();
+      PreparedStatement sql = dbConnPool.run(c -> c.prepareStatement(SQL_SELECT_ALL));
+      sql.setString(1, catalogName);
+      sql.setString(2, JdbcUtil.namespaceToString(namespace));
+      ResultSet rs = sql.executeQuery();
```

Comment:
Could this use `try-with-resources` instead of calling close later?

Reply:
I fixed this in the version above. That uses a try-with-resources block for the prepared statement so that all of the resources are released before releasing the connection back to the pool.